# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: media/mtransport/mediapacket.h
# Commit: c414d82a5325
# Full Hash: c414d82a532508523d40ce63111cb30b2a1076be
# Author: Byron Campen [:bwc] <docfaraday@gmail.com>
# Date: 2019-02-22 05:26:22
# Regressor Bug: 1521879
# File Overlap Count: 1
# Description:
#   Bug 1521879 - Part 1: IPC-based MediaTransport implementation r=mjf
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D17273
# ==============================================================================

diff -r 3a182b05e2c1 -r c414d82a5325 media/mtransport/mediapacket.h
--- a/media/mtransport/mediapacket.h	Thu Feb 21 19:01:32 2019 +0200
+++ b/media/mtransport/mediapacket.h	Thu Feb 21 16:42:12 2019 +0000
@@ -12,6 +12,12 @@
 #include "mozilla/UniquePtr.h"
 #include "mozilla/Maybe.h"
 
+class PickleIterator;
+
+namespace IPC {
+class Message;
+}
+
 namespace mozilla {
 
 // TODO: It might be worthwhile to teach this class how to "borrow" a buffer.
@@ -20,6 +26,7 @@
  public:
   MediaPacket() = default;
   MediaPacket(MediaPacket&& orig) = default;
+  MediaPacket(const MediaPacket& orig);
 
   // Takes ownership of the passed-in data
   void Take(UniquePtr<uint8_t[]>&& data, size_t len, size_t capacity = 0) {
@@ -35,6 +42,9 @@
     data_.reset();
     len_ = 0;
     capacity_ = 0;
+    encrypted_data_.reset();
+    encrypted_len_ = 0;
+    sdp_level_.reset();
   }
 
   // Copies the passed-in data
@@ -68,6 +78,9 @@
 
   Type type() const { return type_; }
 
+  void Serialize(IPC::Message* aMsg) const;
+  bool Deserialize(const IPC::Message* aMsg, PickleIterator* aIter);
+
  private:
   UniquePtr<uint8_t[]> data_;
   size_t len_ = 0;
@@ -80,4 +93,21 @@
   Type type_ = UNCLASSIFIED;
 };
 }  // namespace mozilla
+
+namespace IPC {
+template <typename>
+struct ParamTraits;
+
+template <>
+struct ParamTraits<mozilla::MediaPacket> {
+  static void Write(Message* aMsg, const mozilla::MediaPacket& aParam) {
+    aParam.Serialize(aMsg);
+  }
+
+  static bool Read(const Message* aMsg, PickleIterator* aIter,
+                   mozilla::MediaPacket* aResult) {
+    return aResult->Deserialize(aMsg, aIter);
+  }
+};
+}  // namespace IPC
 #endif  // mediapacket_h__